akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala

```diff
@@ -204,6 +204,18 @@ object Source {
     })
 
   /**
+   * Create [[Source]] that will continually produce given elements in specified order.
+   *
+   * Start a new 'cycled' `Source` from the given elements. The producer stream of elements
+   * will continue infinitely by repeating the same sequence of elements provided during construction.
+   */
+  def cycle[T](f: () ⇒ Iterator[T]): Source[T, NotUsed] = {
+    require(f().nonEmpty, "Provided iterator should not be empty")
```

Comment:
@kelebra I don't think it's expected that `f` is applied here, we should not do that before materialization, i.e. the check should be later and fail the stream instead